dfget/core/core.go

```diff
@@ -189,8 +190,36 @@ func downloadFile(cfg *config.Config, supernodeAPI api.SupernodeAPI,
 	reportFinishedTask(cfg, getter)
 
 	os.Remove(cfg.RV.TempTarget)
-	logrus.Infof("download %s cost:%.3fs length:%d reason:%d",
-		success, time.Since(cfg.StartTime).Seconds(), cfg.RV.FileLength, cfg.BackSourceReason)
+	downloadTime := time.Since(cfg.StartTime).Seconds()
+	if cfg.UploadMetrics {
+		req := &types.TaskMetricsRequest{
+			BacksourceReason: strconv.Itoa(cfg.BackSourceReason),
+			IP:               cfg.RV.LocalIP,
+			CID:              cfg.RV.Cid,
+			CallSystem:       cfg.CallSystem,
+			Duration:         downloadTime,
+			FileLength:       cfg.RV.FileLength,
+			Port:             int32(cfg.RV.PeerPort),
+			Success:          success,
+			TaskID:           result.TaskID,
+		}
+		for _, node := range cfg.Node {
+			resp, err := supernodeAPI.ReportMetrics(node, req)
+			if err != nil {
+				logrus.Error(err)
+			} else if resp.IsSuccess() {
+				break
+			}
+		}
+	}
+
+	if success {
+		logrus.Infof("download SUCCESS cost:%.3fs length:%d",
+			time.Since(cfg.StartTime).Seconds(), cfg.RV.FileLength)
+	} else {
+		logrus.Infof("download FAIL cost:%.3fs length:%d reason:%d",
+			time.Since(cfg.StartTime).Seconds(), cfg.RV.FileLength, cfg.BackSourceReason)
+	}
 	return err
```

Comment:
`return err` is unreasonable, I think. 
Even if the report metric action fails, the function`downloadFile ` should return nil, right?

Reply:
No. This error stands for the error from function `DoDownloadTimeout`, we must return this error if that function failed because it relates to the main downloading logic. 
As for the error of reporting metrics, it will never be returned. We just log it.

Reply:
Then I suggest that encapsulate the code block you added into a single function. That brings much more certainty.